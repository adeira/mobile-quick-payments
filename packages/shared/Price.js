// @flow

import * as React from 'react';
import Translation from 'quick-payments-translations';

const SupportedLocales = {
  // TODO
  en: true,
};

type Props = {|
  // TODO:
  // Amount should be integer (smallest currency unit) so for example
  // cents for EUR (100 cents = 1 EUR) or 100 for Â¥100 (zero-decimal currency).
  +amount: ?number,
  +currency: ?string, // TODO: SupportedCurrencies (?)
  +locale?: $Keys<typeof SupportedLocales>,
|};

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/NumberFormat
 */
export default ({amount, currency, locale = 'en'}: Props) => {
  if (amount == null || currency == null) {
    return null;
  }

  // $FlowIssue: https://github.com/facebook/flow/issues/2801
  const price = Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
  }).format(amount);

  return <Translation passThrough={price} />;
};
