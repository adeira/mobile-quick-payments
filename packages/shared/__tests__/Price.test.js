// @flow

import * as React from 'react';
import Renderer from 'react-test-renderer';

import Price from '../Price';

it('renders Price component as expected', () => {
  expect(
    Renderer.create(<Price amount={0} currency="MXN" />),
  ).toMatchSnapshot();

  expect(
    Renderer.create(<Price amount={10.456} currency="MXN" />),
  ).toMatchSnapshot();

  expect(
    Renderer.create(<Price amount={1e6} currency="MXN" />),
  ).toMatchSnapshot();
});

it('handles falsy values', () => {
  expect(
    Renderer.create(<Price amount={null} currency="MXN" />).toJSON(),
  ).toBeNull();

  expect(
    Renderer.create(<Price amount={10} currency={null} />).toJSON(),
  ).toBeNull();

  expect(
    Renderer.create(<Price amount={null} currency={null} />).toJSON(),
  ).toBeNull();
});
