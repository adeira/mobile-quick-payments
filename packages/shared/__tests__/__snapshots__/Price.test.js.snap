// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Price component as expected 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  MX$0.00
</Text>
`;

exports[`renders Price component as expected 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  MX$10.46
</Text>
`;

exports[`renders Price component as expected 3`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  MX$1,000,000.00
</Text>
`;
